src/Microsoft.AspNetCore.Mvc.Razor/ViewComponentTagHelpers/ViewComponentTagHelperDescriptorFactory.cs

```diff
@@ -0,0 +1,162 @@
+            return tagHelperDescriptors;
+        }
+
+        /*
+        private IEnumerable<TagHelperDescriptor> CreateDescriptors(
+            IEnumerable<ViewComponentDescriptor> viewComponentDescriptors) =>
+            viewComponentDescriptors.Select(viewComponentDescriptor => CreateDescriptor(viewComponentDescriptor));
+            */
+
+        private IEnumerable<TagHelperDescriptor> CreateDescriptors(IEnumerable<ViewComponentDescriptor> viewComponentDescriptors)
+        {
+            var tagHelperDescriptors = new List<TagHelperDescriptor>();
+
+            foreach (var viewComponent in viewComponentDescriptors)
+            {
+                var descriptor = CreateDescriptor(viewComponent);
+                tagHelperDescriptors.Add(descriptor);
+            }
+
+            return tagHelperDescriptors;
+        }
+        private TagHelperDescriptor CreateDescriptor(ViewComponentDescriptor viewComponentDescriptor)
+        {
+            var assemblyName = viewComponentDescriptor.TypeInfo.Assembly.GetName().Name;
+            var tagName = GetTagName(viewComponentDescriptor);
+            var typeName = GetTypeName(viewComponentDescriptor);
+
+            var tagHelperDescriptor = new TagHelperDescriptor
+            {
+                TagName = tagName,
+                TypeName = typeName,
+                AssemblyName = assemblyName,
+                TagStructure = TagStructure.NormalOrSelfClosing,
+            };
+
+            SetAttributeDescriptors(viewComponentDescriptor, tagHelperDescriptor);
+            SetRequiredAttributeDescriptors(viewComponentDescriptor, tagHelperDescriptor);
+
+            tagHelperDescriptor.PropertyBag.Add(
+                ViewComponentTagHelperDescriptorConventions.ViewComponentNameKey, viewComponentDescriptor.ShortName);
+            tagHelperDescriptor.PropertyBag.Add(
+                ViewComponentTagHelperDescriptorConventions.ViewComponentTagHelperNameKey,
+                GetTypeName(viewComponentDescriptor));
+
+            return tagHelperDescriptor;
+        }
+
+        private void SetAttributeDescriptors(ViewComponentDescriptor viewComponentDescriptor,
+            TagHelperDescriptor tagHelperDescriptor)
+        {
+            var methodParameters = viewComponentDescriptor.MethodInfo.GetParameters();
+            var attributeDescriptors = new List<TagHelperAttributeDescriptor>();
+
+            foreach (var parameter in methodParameters)
+            {
+                var lowerKebabName = TagHelperDescriptorFactory.ToHtmlCase(parameter.Name);
+                var descriptor = new TagHelperAttributeDescriptor
+                {
+                    Name = lowerKebabName,
+                    PropertyName = parameter.Name,
+                    TypeName = parameter.ParameterType.FullName
+                };
+
+                var tagHelperType = Type.GetType(descriptor.TypeName);
+                if (tagHelperType.Equals(typeof(string)))
```

Comment:
Looking at the Razor code you actually shouldn't need to do this assuming your TypeName set is accurate. See [here](https://github.com/aspnet/Razor/blob/127e095370e24049faf62e753c89b9b74566e10d/src/Microsoft.AspNetCore.Razor/Compilation/TagHelpers/TagHelperAttributeDescriptor.cs#L125).